src/common/src/util/compress.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+
+/// This function compresses sequential repeated data in a vector. The compression result contains
+/// two vectors, one for the last indices of sequential repeated elements, and another for the
+/// repeated data. For example, [14, 14, 14, 27, 27] will be compressed to [2, 4], [14, 27].
+pub fn compress_data<T>(original_data: &[T]) -> (Vec<u64>, Vec<T>)
+where
+    T: PartialEq + Copy,
+{
+    let mut original_indices = Vec::new();
+    let mut data = Vec::new();
+
+    for i in 1..original_data.len() {
+        if original_data[i - 1] != original_data[i] {
+            original_indices.push(i as u64 - 1);
+            data.push(original_data[i - 1]);
+        }
+    }
+
+    if let Some(&last) = original_data.last() {
+        original_indices.push(original_data.len() as u64 - 1);
+        data.push(last);
+    }
+
+    (original_indices, data)
+}
+
+/// Works in a reversed way as `compress_data`.
+pub fn decompress_data<T>(original_indices: Vec<u64>, data: Vec<T>) -> Vec<T>
```

Comment:
Seems we only need the slice here?
```suggestion
pub fn decompress_data<T>(original_indices: &[u64], data: &[T]) -> Vec<T>
```